### PR TITLE
Implement the realtime multiplayer ready button

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/11206

https://user-images.githubusercontent.com/1329837/102645877-7fd56480-41a6-11eb-8566-1255eb8a600e.mp4

### CODE REVIEW

File: osu.Game/Screens/Multi/Timeshift/TimeshiftReadyButton.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            BackgroundColour = colours.Green;
+            Triangles.ColourDark = colours.Green;
+            Triangles.ColourLight = colours.GreenLight;
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            Enabled.Value = endDate.Value == null || DateTimeOffset.UtcNow.AddSeconds(30).AddMilliseconds(GameBeatmap.Value.Track.Length) < endDate.Value;
```
Comment (bdach) [inline osu.Game/Screens/Multi/Timeshift/TimeshiftReadyButton.cs:35]: This `endDate` null check wasn't there before - are there eventually going to be timeshift lobbies without an end date?

File: osu.Game.Tests/Visual/RealtimeMultiplayer/TestSceneRealtimeReadyButton.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+            addClickButtonStep();
+            AddAssert("match started", () => Client.Room?.Users[0].State == MultiplayerUserState.WaitingForLoad);
+        }
+
+        [Test]
+        public void TestBecomeHostWhileReady()
+        {
+            addClickButtonStep();
+            AddStep("make user host", () => Client.TransferHost(Client.Room?.Users[0].UserID ?? 0));
```
Comment (bdach) [inline osu.Game.Tests/Visual/RealtimeMultiplayer/TestSceneRealtimeReadyButton.cs:101]: There's actually no second user in this room, so this step has no effect and this test doesn't test what it says it does. The current user already has host.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
Not entirely sure about the UX of this. The one particular case I have in mind is readying up when you have host.

* If everyone else has readied up, only the button text changes, the colour doesn't. So the click sorta doesn't seem to have any effect unless you spot the changing copy.
* If there's someone that hasn't readied up, the button changes colour and text.

My main issue is that the button change is too easy to miss. What I would propose instead:

* When not ready, the button is a neutral colour, not green (let's say gray); text is "Ready"
* When readying up:
	* If not host, then the button goes green; text is "Waiting for host..."
	* If host and not all players have readied up, the button goes yellow as currently; not sure about the text, maybe something like "Start game (ready players only)"
	* If host and all players have readied up, the button goes green, with text "All ready!"

**Comment by peppy** [comment]
I discussed this in voice with @smoogipoo, but optimally either on the button text or next to it, it would show `Start game (1 / 9 ready)` or similar.

File: osu.Game/Screens/Multi/Timeshift/TimeshiftReadyButton.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game/Screens/Multi/Timeshift/TimeshiftReadyButton.cs:35]: This is actually an error. Good catch.

**Review by peppy — APPROVED** [review]
👍🏻

**Pull request merged**